File: src/ERGODICITYCostMetric.cpp
#include "ERGODICITYCostMetric.h"
#include <cmath>

ERGODICITYCostMetric::ERGODICITYCostMetric()
	: k_resolution(0), agents_expertnumber(0)
{
}

ERGODICITYCostMetric::ERGODICITYCostMetric(int k_r) 
	: k_resolution(k_r), agents_expertnumber(0)
{

}

/*
输入参数：
xp1：（完整）一维数组数据的指针
xp2：（完整）一维数组数据的指针
phi_r：指针
*/
double ERGODICITYCostMetric::FitMetric(double* xp1, double* xp2, double** phi_r) {
	int n_dimension = 2;
	double result_epsilon = 0;
	double GAMMA_r;

	// xp1,xp2要平移
	for (int rp1 = 0; rp1 < k_resolution + 1; rp1++) {
		for (int rp2 = 0; rp2 < k_resolution + 1; rp2++) {
			GAMMA_r = pow((1 + std::pow(rp1, 2) + std::pow(rp2, 2)), -(n_dimension + 1) / 2);
			double sigma_r = FitFourierBasis(rp1, rp2, xp1, xp2);
			result_epsilon = GAMMA_r * pow(sigma_r - phi_r[rp1][rp2], 2) + result_epsilon;
		}
	}
	return result_epsilon;
}

double** ERGODICITYCostMetric::FitSigmak(double* xp1, double* xp2) {
	double** Sigmak = new double* [k_resolution + 1];
	//double* a[k_resolution]不可以？
	for (int i = 0; i < k_resolution + 1; i++) {
		Sigmak[i] = new double[k_resolution + 1];
	}

	for (int rp1 = 0; rp1 < k_resolution + 1; rp1++) {
		for (int rp2 = 0; rp2 < k_resolution + 1; rp2++) {
			Sigmak[rp1][rp2] = FitFourierBasis(rp1, rp2, xp1, xp2);
		}
	}

	return Sigmak;
}

File: example/matlab/ERGODICITYCostMetric.m
classdef ERGODICITYCostMetric < FOURIERBasis
    %ERGODICITYCOSTMETRIC 此处显示有关此类的摘要
    %   此处显示详细说明
    
    properties (SetAccess = public)
        k_resolution;
        agents_expertnumber;
    end
    
    methods
        function obj = ERGODICITYCostMetric(k_r)
            if nargin>0
                obj.k_resolution = k_r;
            end
        end
        
        function result_epsilon = FitMetric(obj, xp1, xp2, Phik)
            n_dimension = 2;
            result_epsilon = 0;

            for rp1 = 0:obj.k_resolution
                for rp2 = 0:obj.k_resolution
                    GAMMA_r = ((1 + norm([rp1, rp2])^2)^( -(n_dimension+1)/2));
                    sigma_r = FitFourierBasis(obj, rp1, rp2, xp1, xp2);
                    result_epsilon = GAMMA_r * norm(sigma_r - Phik(rp1+1, rp2+1))^2 + result_epsilon;
                    
                    clc;
                    fprintf('[Ergodicity Metric Calculation]: ');
                    fprintf('k1 = %d, k2 = %d \n', rp1, rp2);
                end
            end
        end

        function Sigmak = FitSigmak(obj, xp1, xp2)
            Sigmak = zeros(obj.k_resolution+1);
            for rp1 = 0:obj.k_resolution
                for rp2 = 0:obj.k_resolution
                    Sigmak(rp1+1, rp2+1) = FitFourierBasis(obj, rp1, rp2, xp1, xp2);

                    clc;
                    fprintf('[FourierBasis Calculation]: ');
                    fprintf('k1 = %d, k2 = %d \n', rp1, rp2);
                end
            end

        end
    end
end

